src/glum/_glm.py

```diff
@@ -1234,6 +1326,11 @@ def linear_predictor(
             Sets the alpha(s) to use in case ``alpha_search`` is ``True``.
             Incompatible with ``alpha_index`` (see above).
 
+        context : Optional[Union[int, Mapping[str, Any]]], default=0
+            The context to use for evaluating the formula. If an integer, the
+            context is taken from the stack frame of the caller at the given
+            depth. If a dict, it is directly used as the context.
```

Comment:
Context can be used for custom transformations inside the formula, see _Adding transforms to the evaluation context_ [here](https://github.com/matthewwardrop/formulaic/blob/main/docsite/docs/guides/transforms.ipynb). I agree that this docstring does not really explain what a context does, will try to come up with a better one.

Reply:
That is a fair point. I like the idea of exposing the context only at instantiation. @stanmart what do you think?

Reply:
The issue is that we don't store the context at instantiation, so we need to resupply it (either explicitly or capture from the calling scope) when using prediction-type functions. We are consistent with formulaic in this regard. We could be fancier, but

- We cannot capture everything as the model objects would be huge.
- We could try to figure out which objects the formula needs, and store only those in the model object, but it is tricky. For example, we might see that a formula depends on some function `f` from the local scope. But we would not know if e.g. `f` internally uses `g` also from the local scope. In this case, we would only store `f` in the saved context, and run into a `NameError` on prediction. There are probably other such edge cases, too.

So I think it's better not to try, and rely on the user making sure that the relevant variables and functions are available at prediction time (either explicitly supplied via the `context` dictionary, or being available in the calling scope.

Reply:
> Why would the objects be huge? Why do we need anything other than the initial formula?

Nope, you're right: if we only stored object from the calling scope that the formula actually uses (most likely some functions or constants), that would not be large. With the first point, I only meant that we cannot store everything indiscriminately from the calling scope, as there are probably some large objects there (like, a dataset on which the model is fit).

The other option would be to figure out which objects the formula uses, and store only those in some `_context` attribute. This, however, is also problematic. Let's look at the following example:

```python
def g(x): return x ** 2  # some pre-processing substep
def f(x): return g(x) + 2  # some pre-processing function

df = pd.DataFrame({"x": [1, 2, 3, 4], "y": [1, 1, 2, 2]})
model = glum.GeneralizedLinearModel(formula="y ~ f(x)").fit(df)
```

If we do not store the relevant part of the context, and the prediction function cannot re-capture it either, then we run into the following problem:
```python
>>> model.predict(df, context={})
FactorEvaluationError: Unable to evaluate factor `f(x)`. [NameError: name 'f' is not defined]
```

Let's say we find a way to figure out what objects we need, and save them in some model attribute:
```python
model._context = {"f": f}
```

This approach has two problems. First, and most importantly, (as I realized when playing around with this example) functions are not pickleable, so the model would not be pickleable either:
```python
>>> with open("model.pkl", "wb") as file: pickle.dump(model, file)
# new interpreter:
>>> with open("model.pkl", "rb") as file: model = pickle.load(file)
AttributeError: Can't get attribute 'f' on <module '__main__'>

```

 The other issue is that it is hard to capture everything. For example, in the example above, based on the formula it seems like we are saving everything we need, but that's not the case. So even if we could pickle the function (or use `dill` to do it), we would run into the following issue:
```python
>>> del f, g
>>> model.predict(df, context=model._context)
FactorEvaluationError: Unable to evaluate factor `f(x)`. [NameError: name 'g' is not defined]
```

I cannot think of a good way to handle either of these issues, so I think it makes sense to make the user responsible for having everything that the formula relies on be available during prediction, be it within the dataset (i.e. columns) or outside of it.

Reply:
Sorry to push you further, but what would happen if `f` and `g` were module functions (e.g., `np.log`) instead of lambda functions? Would we be able to pickle all of the information we need? I'd rather require pickle-compatible compatible functions than allow or need users to set the context in each method call, which is likely to be inconvenient and error-prone. (Say, if I want to load someone else's model from MLflow, I'd have to know the context to be able to use it.)

Reply:
I see now that adding the initial context to the model would not be a good idea.

We could also consider removing the context altogether (see also #731). However, when building exploratory models in interactive mode, I found the possibility of quickly defining functions and adding them to the formula via context quite handy. This interactive mode is also the main use for the formula interface IMO. For production models, I would just stick to sklearn-pipelines.

What about keeping the context arguments as is, but changing the default to an empty dict? Then, users would need to opt in to the "proceed-at-your-own-risk"-mode with local context, and the default would be safer.

Reply:
> For production models, I would just stick to sklearn-pipelines.
Agreed, the main use here is the exploratory modeling. I would not upload any model for MLFlow (or use it in production) if it relies on context in any way. When doing stuff interactively, however, there is no need to set the context explicitly: users can rely on glum grabbing stuff from the calling scope.

> what would happen if f and g were module functions (e.g., np.log) instead of lambda functions?
`numpy` functions are a special case as the numpy namespace is automatically added  by formulaic, so those work out of the box. I'll need to do some testing to see how it works with other module functions.

> What about keeping the context arguments as is, except for changing the default to an empty dict? Then, users would need to opt in to the "proceed-at-your-own-risk"-mode with local context, and the default would be safer.

It might work if we special case some argument for grabbing the calling scope (e.g. some string). Otherwise, passing all object one needs is tedious and kinda defeats the purpose for ease of analysis.

Reply:
> It might work if we special case some argument for grabbing the calling scope (e.g. some string)

Wouldn't using `context=0` accomplish that goal? Passing that as an additional argument to utilize local context doesn't appear overly burdensome.

Reply:
Ah yes, you're right.

Reply:
Why do we need to allow users to set the context though? Couldn't we just try to get it from the "calling scope" and, if that fails, raise?

Reply:
> Couldn't we just try to get it from the "calling scope" and, if that fails, raise?

Unless I misunderstand you, that's the current default.

> Why do we need to allow users to set the context though?

I guess the idea is that we want to let users disable grabbing things from the calling scope (`context={}`), and also to let them be very explicit in what they use from outside of the dataset if they want. It brings no extra complexity, because we need to pass contexts around internally between the various prediction functions anyways. It also mirrors [formulaic's interface](https://github.com/matthewwardrop/formulaic/blob/c064ed3be5bbf2419d5d7aac4bf812f42e484d63/formulaic/sugar.py#L9-L15), on which this feature is built.

I'm okay with either default (`context=0` or `context={}`), but I think we should keep the argument as both behaviors are valuable in certain use cases.

Reply:
The current default with `context=0` would be what you describe if I understood correctly. We could also remove the `context` argument and instead expose a Boolean argument `use_calling_scope` or so that can represent `context=0` cand `context={}`. Two reasons against this, though, are:
- There might be use cases in which one wants to fit a model inside a function or similar and pass the context from outside that function.
- We would deviate quite a lot from from formulaic's API.

Reply:
I changed the default to `context=None`. Feel free to revert or edit directly, but I think this is a good compromise.

Reply:
LGTM. Perhaps we could mention in the docstring that "use `context=0` to make the calling scope available"? Just that it's clearer for less techy users.

Also, should we change the default in tabmat too, so that we are consistent across these packages?